### DISCUSSION TITLE
Explore optimizations of AST serialization (preamble)

### DISCUSSION THREAD

**Discussion by sam-mccall** [clangd/clangd#957] [Ideas]
@ilya-biryukov found that in our unit tests something like 40% of buildPreamble CPU is serializing the structures (into memory). This is likely to be also high, though with a different profile, on real large files.

There were some suspicious code patterns, like byte-by-byte copies, and (overall) 10% of CPU time sorting identifier tables for stability.

There's probably potential for serious wins here, preamble builds are ~80% of typical clangd CPU (outside of background indexing).

**Comment by kadircet** [comment]
converting to a discussion as there are no concrete actions here, but we should definitely take a look at this.
